codegen/src/api/runtime_apis.rs

```diff
@@ -0,0 +1,169 @@
+    let docs = &api.docs;
+    let docs: TokenStream2 = should_gen_docs
+        .then_some(quote! { #( #[doc = #docs ] )* })
+        .unwrap_or_default();
+
+    let structs_and_methods: Vec<_> = api.methods.iter().map(|method| {
+        let method_name = format_ident!("{}", method.name);
+
+        // Runtime function name is `TraitName_MethodName`.
+        let runtime_fn_name = format!("{}_{}", trait_name, method_name);
+        let docs = &method.docs;
+        let docs: TokenStream2 = should_gen_docs
+            .then_some(quote! { #( #[doc = #docs ] )* })
+            .unwrap_or_default();
+
+        let inputs: Vec<_> = method.inputs.iter().map(|input| {
+            let name = format_ident!("{}", &input.name);
+            let ty = type_gen.resolve_type_path(input.ty.id);
+
+            let param = quote!(#name: #ty);
+            (param, name)
+        }).collect();
+
+        let params = inputs.iter().map(|(param, _)| param);
+        let param_names = inputs.iter().map(|(_, name)| name);
+
+        // From the method metadata generate a structure that holds
+        // all parameter types. This structure is used with metadata
+        // to encode parameters to the call via `encode_as_fields_to`.
+        let derives = type_gen.default_derives();
+        let struct_name = format_ident!("{}", method.name.to_upper_camel_case());
+        let struct_params = params.clone();
+        let struct_input = quote!(
+            #derives
+            pub struct #struct_name {
+                #( pub #struct_params, )*
+            }
+        );
+
+        let output = type_gen.resolve_type_path(method.output.id);
+
+        let Ok(call_hash) =
+            subxt_metadata::get_runtime_api_hash(metadata, trait_name, &method.name) else {
+                return Err(CodegenError::MissingRuntimeApiMetadata(
+                    trait_name.into(),
+                    method.name.clone(),
+                ))
+            };
+
+        let method = quote!(
+            #docs
+            pub fn #method_name(&self, #( #params, )* ) -> #crate_path::runtime_api::Payload<inputs::#struct_name, #output> {
+                #crate_path::runtime_api::Payload::new_static(
+                    #runtime_fn_name,
+                    inputs::#struct_name { #( #param_names, )* },
+                    [#(#call_hash,)*],
+                )
+            }
+        );
+
+        Ok((struct_input, method))
+    }).collect::<Result<_, _>>()?;
+
+    let trait_name = format_ident!("{}", trait_name);
+
+    let structs = structs_and_methods.iter().map(|(struct_, _)| struct_);
+    let methods = structs_and_methods.iter().map(|(_, method)| method);
+
+    let runtime_api = quote!(
+        pub mod #trait_name_snake {
+            use super::root_mod;
+            use super::#types_mod_ident;
+
+            #docs
+            pub struct #trait_name;
+
+            impl #trait_name {
+                #( #methods )*
+            }
+
+            pub mod inputs {
```

Comment:
I quite like this `inputs` mod; maybe we should do the same for calls (and storage? and whatever)?

Not a blocker or anythign though :)

Reply:
After looking at the code a bit, leaving here my thoughts on this:

- For Calls we do something similar, starting from the Call enum we generate the structures of the payloads.
From what I see those structs are used as parameters for the payload itself, as a result, not necessarily as an input: `subxt::tx::Payload<Send>`. 
We could rename the `inputs` mod to `types` to provide a more generic naming that could be used here as well.

```rust
pub struct AddMemo {
    pub index: runtime_types::polkadot_parachain::primitives::Id,
    pub memo: ::std::vec::Vec<::core::primitive::u8>,
}

pub struct TransactionApi;
impl TransactionApi {
    pub fn add_memo(
        &self,
        index: runtime_types::polkadot_parachain::primitives::Id,
        memo: ::std::vec::Vec<::core::primitive::u8>,
    ) -> ::subxt::tx::Payload<AddMemo> {
        ::subxt::tx::Payload::new_static(
            "Crowdloan",
            "add_memo",
            AddMemo { index, memo }, ...
```
We could place that `AddMemo` into `pub mod types` for example.

- We use a similar approach to Calls for generating Events, because we registered the event type in the metadata as an enum.
The enum module generated for a pallet only contains the generated structs; as opposed to Calls that also generate the `struct TransactionApi`. Since there's nothing that could collide, I would leave Events untouched.

- The return type of Constants is already generated by the `TypeGenerator` (because that specific type is already present / registered in the metadata; as opposed to the `Call` enum or vector of type IDs from which we build structs).

- For storage entries, we use the return type directly similar to Constants. For the function parameters however, we pass the parameters directly to the `subxt::storage::address::make_static_storage_map_key`. And since I believe we don't have enough type information to name the parameters, we use `_0`, `_1` etc. 

```rust
pub struct StorageApi;
impl StorageApi {
pub fn queries(
      &self,
      _0: impl ::std::borrow::Borrow<::core::primitive::u64>,
  ) -> ::subxt::storage::address::Address<...
  > {
      ::subxt::storage::address::Address::new_static(
          "XcmPallet", "Queries",
          vec![::subxt::storage::address::make_static_storage_map_key(
              _0.borrow(),
          )],
      )
  }
``` 

And the `make_static_storage_map_key` returns:
https://github.com/paritytech/subxt/blob/b597b5d331e0385ba7a2973baeaf505cd405e104/subxt/src/storage/storage_address.rs#L213-L215

I'm wondering if we could generate structs for those storage calls tho. 
I think we didn't do that because we'd need to hash and encode independently each field of the struct, and that might turn out to be a bit tricky.

TLDR; will change the input to types and do a similar change for calls. Let me know what you think of it 🙏

Reply:
The example for calls that you gave:

```rust
impl TransactionApi {
    pub fn add_memo(
        &self,
        index: runtime_types::polkadot_parachain::primitives::Id,
        memo: ::std::vec::Vec<::core::primitive::u8>,
    ) -> ::subxt::tx::Payload<AddMemo> {
        ::subxt::tx::Payload::new_static(
            "Crowdloan",
            "add_memo",
            AddMemo { index, memo }, ...
```

Isn't that identical to the runtime API usage of structs, ie `AddMemo` is an input type to the payload just like it's an input on the runtime apis?

That said, calling it `types` instead of `inputs` I'd have no issue at all with; it's a bit more generic in case it's a useful module name to use in other places :)

Not a big deal either way; happy with whatyou you think best here!

Reply:
> Isn't that identical to the runtime API usage of structs, ie AddMemo is an input type to the payload just like it's an input on the runtime apis?

Yep, they are identical, we generate them almost the same way. 

Thanks! Will go ahead and place them under the `types` module